velox/vector/VariantToVector.cpp

```diff
@@ -77,4 +77,165 @@ ArrayVectorPtr variantArrayToVector(
       pool);
 }
 
+// Structure used to keep check on data type of
+// the complex vector, along with the number and data
+// types of constituent elements. Attributes elementCount
+// and elementInserted keeps the length of the vector, and
+// the number of elements inserted during the operation
+// respectively.
+struct NestedTypeCounter {
+  TypeKind kind = TypeKind::UNKNOWN;
+  vector_size_t elementCount = 0;
+  vector_size_t elementsInserted = 0;
+  std::vector<NestedTypeCounter> children;
+};
+
+static void validateOrSetType(NestedTypeCounter& counter, const variant& v) {
+  if (v.isNull()) // If a variant is null, it can be UNKNOWN, INVALID, or
+                  // matching our expected kind
+  {
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
+        v.kind() != counter.kind) {
+      throw std::invalid_argument("Variant was of an unexpected kind");
+    }
+    return;
+  } else {
+    if (v.kind() == TypeKind::UNKNOWN || v.kind() == TypeKind::INVALID) {
+      throw std::invalid_argument(
+          "Non-null variant has unknown or invalid kind");
+    }
+    if (v.kind() != counter.kind) {
+      if (counter.kind == TypeKind::UNKNOWN) {
+        counter.kind = v.kind();
+      } else {
+        throw std::invalid_argument("Variant kind doesn't match predecessors");
+      }
+    }
+  }
+}
+
+static void computeElementsCountsAndType(
+    NestedTypeCounter& counter,
+    const variant& v) {
+  validateOrSetType(counter, v);
+  counter.elementCount += 1;
+  if (v.isNull()) {
```

Comment:
If v is null, shouldnt we save where its null at so we can mark the appropriate null bit in the nulls buffer ?